### PR TITLE
Skip wrapping IServer when IIS hosted

### LINKED ISSUE #313: Cannot host CoreWcf in IIS under ASP.NET Core 3.1
I cannot find a way of hosting a CoreWcf HTTP SOAP service in IIS or IISExpress on versions of ASP.NET Core later than 2.1. Please close the bug if support for IIS on later dot net versions is already planned, thanks.

## Repo Steps

In Visual Studio:
* Create a new 'ASP.NET Core Web Application'
* 'ASP.NET Core 3.1' 
* Select the 'Empty' option
* Add the CoreWcf Nuget packages
* Add services.AddServiceModelServices(); in ConfigureServices

## Error raised

The exception 'Application is running inside IIS process but is not configured to use IIS serve' is raised running the code:

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Application is running inside IIS process but is not configured to use IIS server.
  Source=Microsoft.AspNetCore.Server.IIS
  StackTrace:
   at Microsoft.AspNetCore.Server.IIS.Core.IISServerSetupFilter.<>c__DisplayClass2_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.HostFilteringStartupFilter.<>c__DisplayClass0_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.<StartAsync>d__31.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>d__9.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at IisCoreWcfTest.Program.Main(String[] args) in C:\Users\jallderidge\source\repos\CoreWcf_IisHostBugReport\IisCoreWcfTest\Program.cs:line 11

  This exception was originally thrown at this call stack:
    Microsoft.AspNetCore.Server.IIS.Core.IISServerSetupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder)

Microsoft.AspNetCore.HostFilteringStartupFilter.Configure.AnonymousMethod__0(Microsoft.AspNetCore.Builder.IApplicationBuilder)
    Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(System.Threading.CancellationToken)
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
    Microsoft.Extensions.Hosting.Internal.Host.StartAsync(System.Threading.CancellationToken)
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
```

## Example Code

```
    public class Program
    {
        public static void Main(string[] args) => CreateHostBuilder(args).Build().Run();

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });
    }

    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            // adding this line will prevent startup in IIS express
            services.AddServiceModelServices();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
        }
    }
```

An example solution is at [https://github.com/allderidge/CoreWcf_IisHostBugReport](https://github.com/allderidge/CoreWcf_IisHostBugReport)

### PR DESCRIPTION
**Pull request by mconnew**
Also removed accidental dummy IServer interface from CoreWCF.Configuration

Fixes #313

### CODE REVIEW

**Review by davidfowl — DISMISSED** [review]
Why do you wrap at all?

**Comment by mconnew** [comment]
@davidfowl, CoreWCF can't be initialized until we know what ports are being used, so we delay initialization until after the host has started. The problem then is that any config problems can't be detected until after the IServer has started. ASP.NET Core lacks any extensibility points where you can run code and throw an exception after the IServer has started. The only extensibility point after the IServer starts catches and swallows all exceptions. This is the only way to have host.Start throw on service initialization errors. If you fix this gap in the api for aspnetcore, I'll happily remove this in a later version 😆

**Comment by davidfowl** [comment]
You should be able to add an IHostedService and check in StartAsync. Does that not work here?

**Comment by mconnew** [comment]
It looks like tests are failing because of an image change in AzDo and the "netstat -a -n" command no longer works on Ubuntu 20.04. This might require talking to AzDo support as the netstat command is present in Ubuntu 20.04 by default. We're seeing the exception "No such file or directory" referring to the netstat executable. I'm going to do some local testing before contacting support.

**Comment by mconnew** [comment]
@davidfowl, IHostedService.StartAsync runs before the host is started on at least some versions of ASP.NET Core. I did discuss this with Stephen Halter first to make sure I wasn't missing something and he agreed this is the only way to make it work reliably on all current versions of ASP.NET Core.

**Comment by davidfowl** [comment]
I think you can use a hosted service combined maybe with the ApplicationStarted event?

**Comment by mconnew** [comment]
ApplicationStarted event swallows exceptions.

**Comment by davidfowl** [comment]
> ApplicationStarted event swallows exceptions.

Of course it does.... OK then. Carry on with his mega hack 😉

**Comment by mconnew** [comment]
Opened an issue at https://github.com/actions/virtual-environments/issues/2837 for the missing netstat on Ubuntu 20.04. For now I'll mark the test as Windows only.

File: src/CoreWCF.Primitives/src/CoreWCF/Configuration/ServiceModelServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -27,6 +27,12 @@ public static IServiceCollection AddServiceModelServices(this IServiceCollection
                 {
                     if (services[i].ImplementationType != null)
                     {
+                        if (services[i].ImplementationType.FullName.Equals("Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer"))
```
Comment (birojnayak) [inline src/CoreWCF.Primitives/src/CoreWCF/Configuration/ServiceModelServiceCollectionExtensions.cs:30]: let's declare a constant

**Review by birojnayak — APPROVED** [review]
good insights based on above discussion 👍 ... one minor comment, otherwise looks good to check in.

**Pull request merged**